velox/functions/sparksql/fuzzer/SparkQueryRunner.cpp

```diff
@@ -0,0 +1,388 @@
+} // namespace
+
+SparkQueryRunner::SparkQueryRunner(const std::string& coordinatorUri) {
+  std::shared_ptr<grpc::Channel> channel =
+      grpc::CreateChannel(coordinatorUri, grpc::InsecureChannelCredentials());
+  stub_ = spark::connect::SparkConnectService::NewStub(channel);
+}
+
+std::optional<std::string> SparkQueryRunner::toSql(
+    const velox::core::PlanNodePtr& plan) {
+  if (const auto aggregationNode =
+          std::dynamic_pointer_cast<const core::AggregationNode>(plan)) {
+    return toSql(aggregationNode);
+  }
+  if (const auto projectNode =
+          std::dynamic_pointer_cast<const core::ProjectNode>(plan)) {
+    return toSql(projectNode);
+  }
+  VELOX_NYI("Unsupported plan node: {}.", plan->toString());
+}
+
+std::multiset<std::vector<variant>> SparkQueryRunner::execute(
+    const std::string& sql,
+    const std::vector<RowVectorPtr>& input,
+    const RowTypePtr& resultType) {
+  return exec::test::materialize(executeVector(sql, input, resultType));
+}
+
+std::vector<RowVectorPtr> SparkQueryRunner::executeVector(
+    const std::string& sql,
+    const std::vector<RowVectorPtr>& input,
+    const RowTypePtr& resultType) {
+  auto inputType = asRowType(input[0]->type());
+  if (inputType->size() == 0) {
+    // The query doesn't need to read any columns, but it needs to see a
+    // specific number of rows. Make new 'input' as single all-null BIGINT
+    // column with as many rows as original input. This way we'll be able to
+    // create a 'tmp' table will the necessary number of rows.
+    vector_size_t numInput = 0;
+    for (const auto& v : input) {
+      numInput += v->size();
+    }
+
+    auto column = BaseVector::createNullConstant(BIGINT(), numInput, pool());
+    auto rowVector = std::make_shared<RowVector>(
+        pool(),
+        ROW({"x"}, {BIGINT()}),
+        nullptr,
+        numInput,
+        std::vector<VectorPtr>{column});
+    return executeVector(sql, {rowVector}, resultType);
+  }
+
+  // Write the input to a Parquet file.
+  auto tempFile = exec::test::TempFilePath::create();
+  const auto& filePath = tempFile->getPath();
+  auto writerPool = rootPool()->addAggregateChild("writer");
+  writeToFile(filePath, input, writerPool.get());
+
+  // Create temporary view 'tmp' in Spark by reading the generated Parquet file.
+  execute(fmt::format(
+      "CREATE OR REPLACE TEMPORARY VIEW tmp AS (SELECT * from parquet.`file://{}`);",
+      filePath));
+  return execute(sql);
+}
+
+std::vector<RowVectorPtr> SparkQueryRunner::execute(
+    const std::string& content) {
+  auto sql = google::protobuf::Arena::CreateMessage<SQL>(&arena_);
+  sql->set_query(content);
+
+  auto relation = google::protobuf::Arena::CreateMessage<Relation>(&arena_);
+  relation->set_allocated_sql(sql);
+
+  auto plan = google::protobuf::Arena::CreateMessage<Plan>(&arena_);
+  plan->set_allocated_root(relation);
+
+  auto context = google::protobuf::Arena::CreateMessage<UserContext>(&arena_);
+  context->set_user_id(kUserId);
```

Comment:
Curious why are we hard-coding user ID / name? Would it make sense to allow these to be specified in the constructor instead?